### PR TITLE
KAFKA-5520: KIP-171 - Extend Consumer Group Reset Offset for Stream Application - Updated

### PR DESCRIPTION
**Pull request by jeqo**
KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-171+-+Extend+Consumer+Group+Reset+Offset+for+Stream+Application

Merge changes from KIP-198

Ref: https://github.com/apache/kafka/pull/3831

### CODE REVIEW

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Can't this be package private? (add comment `// visible for testing`)

nit: `ts` -> `timestamp`

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: This could throw a NPE. I think we should guard against this, and throw `ParseException` if NPE happens

`ts.split("T")` is redundant and should be extracted into a variable.

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java]: this should be multiple tests (one test method per scenario). Also, use better method names, eg, `shouldAcceptValidDateFormats`, `shouldThrowOnInvalidDateFormat` etc

File: streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java]: nit: add `final` (same below)

File: streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java]: `final`

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Should we guard against NPE? (same blow)

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: nit: add `final`

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: nit: As key and value is accessed, we might want to iterate throw the `.entrySet` instead of `keySet`

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: `final` (3x)

**Review by mjsax — CHANGES_REQUESTED** [review]
Some initial comments. Did not yet have a look at the test (it seems, there is a lot of redundancy in the test though -- could we refactor to share more code?)

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (jeqo) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: Agree. I will reformat this.

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (jeqo) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: Yes, but test is in another package/module. Maybe once it is moved to streams we can improve visibility of this method. I will add the comment.

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (jeqo) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: Agree. I'll reformat.

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (jeqo) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: Yeap, I'll refactor.

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (jeqo) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: Agree.

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (jeqo) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: agree.

File: streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java
Comment (jeqo) [reply streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java]: Agree.

File: streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java
Comment (jeqo) [reply streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java]: Not sure where the final should be. Maybe you're referring to the next comment?

File: streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java
Comment (jeqo) [reply streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java]: Yes, I will add them.

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by asfgit** [comment]
FAILURE
 8076 tests run, 5 skipped, 2 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8076 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8076 tests run, 5 skipped, 0 failed.
--none--

File: streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/ResetIntegrationTest.java]: `} catch (final Exception e) {`

**Comment by asfgit** [comment]
FAILURE
 8076 tests run, 5 skipped, 2 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 7959 tests run, 5 skipped, 1 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8076 tests run, 5 skipped, 0 failed.
--none--

**Comment by mjsax** [comment]
@jeqo We just merged a fix for the failing tests. Can you please rebase. thx.

**Comment by asfgit** [comment]
SUCCESS 
 8080 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 7963 tests run, 5 skipped, 1 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8080 tests run, 5 skipped, 0 failed.
--none--

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. Looks good overall.

However, I think we need to refactor the integration test. There was already some code duplication before this, but not it's even worse. I think, that each new test could be an own class, and the parts that differ can either be passed as parameters of be extracted into abstract methods in the main class that are implemented differently in the test classes.

Having said this, I am wondering if we should switch from a class hierarchy to hand in everything (the abstract method could be refactored as "runner classes" that get executed.

Call for second review @guozhangwang @dguy @bbejeck

**Comment by guozhangwang** [comment]
I only made a pass over the testing classes. My suggestion:

1) We can add a `StreamsResetterTest` for unit testing the tool, with a `MockConsumer` and `MockKafkaAdminClientEnv` to validate that the corresponding `seeks`, `commits`, and `deleteRecords` are called with the expected parameters.

2) For integration tests, we only need to maintain one test as we already have now.

**Comment by bbejeck** [comment]
Thanks for the PR, looks good overall just one minor comment.

The `maybeReset` method is long; maybe we could put the code in `if` statements over 2 or 3 lines in separate methods for readability?

**Comment by jeqo** [comment]
@bbejeck agree. I've fix it. 
@guozhangwang I will give a try and check how it goes.

Thanks for your feedback!

**Comment by jeqo** [comment]
@guozhangwang I've added `StreamsResetterTest` to unit-test `seek` and `removeRecords` functionality, and about integration tests, I have add methods to test it with SSL. 
Let me know if there's anything missing to merge :)

**Comment by guozhangwang** [comment]
@jeqo The unit tests of `StreamsResetterTest` lgtm. For the integration test though, my previous comment is that we do not need to add more under integration tests as they should be well covered. LMK.

**Comment by jeqo** [comment]
@guozhangwang I've removed SSL integration test, and put integration tests in one class. how does it look now?

File: streams/src/test/java/org/apache/kafka/streams/integration/StreamsResetterIntegrationTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/StreamsResetterIntegrationTest.java]: @jeqo Maybe my previous comment was misleading :) I was suggesting that we do not need to add more test cases into the reset integration tests like "from offset", "from date" etc, as for integration test we just check its interaction with the cluster module all worked as expected, so having just one test for "from scratch with intermediate" and "from scratch without intermediate" is good enough. For the added functionality our added unit tests should be well covering it.

Also it seems you have removed the test with SSL config as well? Is that intentional? If not maybe we can just make this integration test a parameterized test, with ssl turned on and off (i.e. when it is turned on broker cluster will have the SSL configs to not accept unauthorized access, and the streams resetter needs the appropriate SSL configs to be able to delete those topics).

cc @mjsax

Comment (jeqo) [reply streams/src/test/java/org/apache/kafka/streams/integration/StreamsResetterIntegrationTest.java]: I think I get it know: I will return tests with SSL config (I thought that was what it was already tested) and remove tests that are already validated by unit-test. There are some methods from `KafkaConsumer` that are not available on unit test like `offsetsForTimes`, I will keep those on integration test.

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/StreamsResetterIntegrationTest.java]: Sounds good! Thanks. The reason we'd like to avoid redundant integration tests with overlapping unit tests is that they would take much more time. If we are not careful the Jenkins builds would take longer and longer to finish, and hence hinders our development pace (we have seen the Streams unit test suite taking from 1 min to about 7 min now on my laptop, and much longer on the Jenkins box).

Comment (jeqo) [reply streams/src/test/java/org/apache/kafka/streams/integration/StreamsResetterIntegrationTest.java]: Make total sense 👍

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java]: If we add the intermediate topic for all cases, I am wondering if this is correct.

I understand that the test is still passing and we just log an error, but should StreamsResetter not return a non-zero return value to indicate that something went wrong?

**Comment by mjsax** [comment]
Thanks for the updates. Still wondering if we should share more code for the individual test methods. Do not insist on if, but might be worth. WDYT @guozhangwang ?

**Comment by guozhangwang** [comment]
I looked at the `StreamsResetterTest` and I think it is well readable, while sharing common code may be less effective, so if you are referring to this test class I actually think it is good as is.

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java]: That's a good point. Maybe we should still add a boolean parameter to indicate if we should include `--intermediate-topics", INTERMEDIATE_USER_TOPIC` into the config lists?

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java]: Yes. And we should make sure that the test fails if we do it wrong. That the test passes atm, indicates an issue in the reset tool. It only returns non-zero exit code if there was an fatal exception and we crash. But we should also return non-zero if anything else went wrong. Atm, we catch, print/log and move on -- we should remember if any error occurred though and only return zero if no error occurred at all.

I would recommend to use different return codes for different errors and document it. Should be part of the public API IMHO.

And add a test to check the return codes :)

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java]: Should this be a separate PR out of the scope of the KIP though?

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java]: Absolutely. It's a bug and we can file a JIRA -- thus, the bug fix can also re-introduce the it-then-else for the intermediate topic here.

**Comment by jeqo** [comment]
@mjsax we could go back to private methods once we move `StreamsResetter` to `streams` module (i.e. merge KIP-222), so it should be a temporary solution to unit-test.

File: streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/AbstractResetIntegrationTest.java]: @jeqo do you want to add back the boolean indicator whether intermediate topic exists and should be removed before we merge this PR?

**Comment by mjsax** [comment]
I just looked into `StreamsResetter`. Wondering why the test passed without the `boolean` flag we just added back. Shouldn't we throw here and die if there is no intermediate topic to be deleted, and finally return `EXIT_CODE_ERROR` ?

```
    private void doDelete(final List<String> topicsToDelete,
                          final KafkaAdminClient adminClient) {
        boolean hasDeleteErrors = false;
        final DeleteTopicsResult deleteTopicsResult = adminClient.deleteTopics(topicsToDelete);
        final Map<String, KafkaFuture<Void>> results = deleteTopicsResult.values();

        for (final Map.Entry<String, KafkaFuture<Void>> entry : results.entrySet()) {
            try {
                entry.getValue().get(30, TimeUnit.SECONDS);
            } catch (Exception e) {
                System.err.println("ERROR: deleting topic " + entry.getKey());
                e.printStackTrace(System.err);
                hasDeleteErrors = true;
            }
        }
        if (hasDeleteErrors) {
            throw new RuntimeException("Encountered an error deleting one or more topics");
        }
    }
```

**Comment by jeqo** [comment]
But `doDelete` is to delete internal topics, not intermediate ones. Intermediates are suppose to move offsets to the end with `maybeResetInputAndSeekToEndIntermediateTopicOffsets`

**Comment by guozhangwang** [comment]
The reason that it does not fail is because in the resetter, we use a client to subscribe to all the topics, and then populate the intermediate topics:

```
client.subscribe(topicsToSubscribe);
            client.poll(1);

            final Set<TopicPartition> partitions = client.assignment();
            final Set<TopicPartition> inputTopicPartitions = new HashSet<>();
            final Set<TopicPartition> intermediateTopicPartitions = new HashSet<>();

            for (final TopicPartition p : partitions) {
                final String topic = p.topic();
                if (isInputTopic(topic)) {
                    inputTopicPartitions.add(p);
                } else if (isIntermediateTopic(topic)) {
                    intermediateTopicPartitions.add(p);
                } else {
                    System.err.println("Skipping invalid partition: " + p);
                }
            }
```

If the intermediate topic does not exist, then the `intermediateTopicPartitions` would be empty, and hence `maybeSeekToEnd` will be a no-op, so no exceptions will be thrown.

Synced with @mjsax offline, and he will file a JIRA to improve on this situation separately.

**Comment by mjsax** [comment]
@jeqo You are right -- I did look at the wrong code. My bad. Makes sense that the test passed.

Did create https://issues.apache.org/jira/browse/KAFKA-6318 as follow up.

This is the issue (unrelated to the KIP):
```
            if (notFoundIntermediateTopics.size() > 0) {
                System.out.println("Following intermediate topics are not found, skipping them");
                for (final String topic : notFoundIntermediateTopics) {
                    System.out.println("Topic:" + topic);
                }
            }
```

**Review by mjsax — APPROVED** [review]
Thanks for the KIP and PR! Great addition!

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk.

@jeqo Thanks for your contribution!

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (tedyu) [inline core/src/main/scala/kafka/tools/StreamsResetter.java:347]: Should this call be governed by !dryRun

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (tedyu) [inline core/src/main/scala/kafka/tools/StreamsResetter.java:391]: These two calls boils down to
```
Fetcher#retrieveOffsetsByTimes(Map<TopicPartition, Long> timestampsToSearch ...)
```
I wonder if they can be combined (to save roundtrips).

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (mjsax) [reply core/src/main/scala/kafka/tools/StreamsResetter.java:347]: It's fine. `seekToEnd()` is lazy and we guard `position()` with `dryRun`-check:

https://github.com/apache/kafka/pull/4159/files/fab095f46803e2f1795b4f128a607d5fdbd51423#diff-52e29464b15979ca6abaf9da60e57f86R309

File: core/src/main/scala/kafka/tools/StreamsResetter.java
Comment (mjsax) [reply core/src/main/scala/kafka/tools/StreamsResetter.java:391]: We want to get `endOffsets()` and `beginningOffsets` for the same set of partitions. A single request cannot get both at once AFAIK.

Also, the reset tool is not considered to be on the "hot code path" -- thus, we don't need to worry about performance too much and apply (unnecessary?) micro optimizations. Just my two cents here.

Comment (tedyu) [reply core/src/main/scala/kafka/tools/StreamsResetter.java:391]: Agreed that such optimization is not needed for this task.

**Comment by mjsax** [comment]
@jeqo I just realized, that we should have updated the docs with this PR. Would you mind to do a follow up PR? Should go here https://kafka.apache.org/10/documentation/streams/upgrade-guide and https://kafka.apache.org/10/documentation/streams/developer-guide/app-reset-tool.html

**Pull request closed without merging**
